File: 2019/day5/python/intcode.py
import copy
from collections import defaultdict


class Intcode:
    """Class containing the intcode computer."""

    def __init__(self, file, input=[], verbose=True, reset=True):
        """Read file and give input to the intcode computer."""
        self.memory = self.read_file(file)
        self.x, self.output = None, None
        self.input = input if isinstance(input, list) else [input]
        self.verbose = verbose
        self.reset = reset
        self.idle = False
        self.n = len(self.memory)
        self.i, self.base = 0, 0

    def read_file(self, file):
        try:
            with open(file) as f:
                row = f.read()
            data = defaultdict(
                int, {i: int(x) for i, x in enumerate(row.split(","))}
            )

            return data

        except IOError:
            print("Cannot find file at: " + file)
            return None

    def parse_opcode(self, op):
        """Opcode parsing - analyzes the opcode string."""
        oper = op[-1]
        mode = "000" + op[:-2]
        # Parse position vs immediate mode for each operation
        j = self.i

        # Relative base check
        b0 = self.base if mode[-1] == "2" else 0
        b1 = self.base if mode[-2] == "2" else 0
        b2 = self.base if mode[-3] == "2" else 0

        xb0 = self.x[b0 + self.x[j + 1]]
        xb1 = self.x[b1 + self.x[j + 2]]

        if oper == "1" or oper == "2" or oper == "7" or oper == "8":
            p1 = self.x[j + 1] if mode[-1] == "1" else xb0
            p2 = self.x[j + 2] if mode[-2] == "1" else xb1
            out = b2 + self.x[j + 3]
            self.i = self.i + 4
            return oper, p1, p2, out
        elif oper == "5" or oper == "6":
            p1 = self.x[j + 1] if mode[-1] == "1" else xb0
            p2 = self.x[j + 2] if mode[-2] == "1" else xb1
            out = self.x[b2 + j + 3]
            self.i = self.i + 3
            return oper, p1, p2, None
        elif oper == "4":
            p = self.x[j + 1] if mode[-1] == "1" else xb0
            self.i = self.i + 2
            return oper, p, None, None
        elif oper == "9":
            base = self.x[j + 1] if mode[-1] == "1" else xb0
            self.i = self.i + 2
            self.base += base
            return None, None, None, None
        else:
            p = b0 + self.x[j + 1]
            self.i = self.i + 2
            return oper, p, None, None

    def operate(self, oper, p1, p2, out):
        """Apply operation on the intcode sequence."""

        p1 = 0 if p1 is None else p1
        p2 = 0 if p2 is None else p2
        out = 0 if out is None else out

        write_op = {
            "1": p1 + p2,
            "2": p1 * p2,
            "7": 1 if p1 < p2 else 0,
            "8": 1 if p1 == p2 else 0,
        }

        increment_op = {
            "5": p2 if p1 != 0 else self.i,
            "6": p2 if p1 == 0 else self.i,
        }

        if oper in write_op:
            self.x[out] = write_op[oper]
        elif oper == "3":
            if not self.input:
                self.input.append(-1)
                self.idle = True

            self.x[p1] = self.input[0]
            self.input.pop(0)
        elif oper == "4":
            self.output = p1
            if self.verbose:
                print(p1)
        elif oper in increment_op:
            self.i = increment_op[oper]

    def __call__(self, input=[]):
        """Call to solve the intcode."""

        self.x = copy.deepcopy(self.memory) if self.reset else self.memory

        if isinstance(input, list):
            if len(input) > 0:
                for inp in input:
                    self.input.append(inp)
        else:
            self.input.append(input)

        while self.i < self.n:

            if self.x[self.i] == 99:
                self.i, self.base = 0, 0
                return True, self.output
            oper, p1, p2, out = self.parse_opcode(str(self.x[self.i]))

            if oper == "3" and self.idle:
                return False, None
            self.operate(oper, p1, p2, out)

            if (not self.reset) and oper == "4":
                return False, self.output

        return self.output

File: 2019/day5/cpp/src/intcode.cpp
#include "intcode.hpp"

#include <iostream>
#include <numeric>

#include "utils.hpp"

using std::tuple;

Intcode::Intcode(bool reset, bool verbose)
    : m_program(),
      m_input(),
      m_iter(0),
      m_base(0),
      m_output(0),
      m_reset(reset),
      m_verbose(verbose),
      m_idle(false) {}

Intcode::Intcode(std::string file_name, std::vector<int64_t> input, bool reset,
                 bool verbose)
    : m_program(),
      m_input(input),
      m_iter(0),
      m_base(0),
      m_output(0),
      m_reset(reset),
      m_verbose(verbose),
      m_idle(false) {
  // Read the program
  std::vector<std::string> file_contents = utils::read_file(file_name);
  auto file_concat = std::accumulate(file_contents.begin(), file_contents.end(),
                                     std::string(""));
  std::vector<int64_t> input_file;
  utils::split_string(file_concat, ",", input_file);
  for (auto &inp : input_file) {
    m_program.push_back(inp);
  }
}

// Parse the incoming operation with its operands
tuple<char, int64_t, int64_t, int64_t> Intcode::parse_opcode(std::string op) {
  // Get the operation
  auto oper = op.back();

  // Extract the mode of the operations - pad with zeros to simplify
  auto mode = "000" + op.substr(0, std::max(int(op.length() - 2), 0));
  auto mode_len = mode.length() - 1;

  // Temporary iterating index
  auto j = m_iter;

  // Set base offsets for the out, p2, and p1
  auto b0 = (mode[mode_len] == '2') ? m_base : 0;
  auto b1 = (mode[mode_len - 1] == '2') ? m_base : 0;
  auto b2 = (mode[mode_len - 2] == '2') ? m_base : 0;

  auto b0x = b0 + m_program[j + 1];
  auto b1x = b1 + m_program[j + 2];

  auto xb0 = (b0x >= int64_t(m_program.size()) || b0x < 0) ? 0 : m_program[b0x];
  auto xb1 = (b1x >= int64_t(m_program.size()) || b1x < 0) ? 0 : m_program[b1x];

  /* Set what the operation parameters will be: either its value or
  value at program address. */
  if (oper == '1' || oper == '2' || oper == '7' || oper == '8') {
    int64_t p1 = (mode[mode_len] == '1') ? m_program[j + 1] : xb0;
    int64_t p2 = (mode[mode_len - 1] == '1') ? m_program[j + 2] : xb1;
    int64_t out = b2 + m_program[j + 3];
    m_iter += 4;
    return std::make_tuple(oper, p1, p2, out);
  } else if (oper == '5' or oper == '6') {
    int64_t p1 = (mode[mode_len] == '1') ? m_program[j + 1] : xb0;
    int64_t p2 = (mode[mode_len - 1] == '1') ? m_program[j + 2] : xb1;
    m_iter += 3;
    return std::make_tuple(oper, p1, p2, -1);
  } else if (oper == '4') {
    auto p = (mode[mode_len] == '1') ? m_program[j + 1] : xb0;
    m_iter += 2;
    return std::make_tuple(oper, p, -1, -1);
  } else if (oper == '9') {
    auto base_ = (mode[mode_len] == '1') ? m_program[j + 1] : xb0;
    m_iter += 2;
    m_base += base_;
    return std::make_tuple(' ', -1, -1, -1);
  } else {
    auto p = b0 + m_program[j + 1];
    m_iter += 2;
    return std::make_tuple(oper, p, -1, -1);
  }
}

// Apply the operations
void Intcode::operate(char oper, int64_t p1, int64_t p2, int64_t out) {
  // If we need to store memory outside our program, resize.
  if (out >= int64_t(m_program.size())) {
    if (oper == '1' || oper == '2' || oper == '7' || oper == '8') {
      m_program.resize(out * 2);
    }
  }

  if (oper == '1') {
    m_program[out] = p1 + p2;
  } else if (oper == '2') {
    m_program[out] = p1 * p2;
  } else if (oper == '3') {
    // If obtained an input instruction, but did not provide one, go idle.
    if (m_input.size() == 0) {
      m_input.push_back(-1);
      m_idle = true;
    }
    if (p1 >= int64_t(m_program.size())) {
      m_program.resize(2 * p1);
    }
    m_program[p1] = m_input.front();
    m_input.erase(m_input.begin());
  } else if (oper == '4') {
    m_output = p1;
    if (m_verbose) {
      std::cout << p1 << "\n";
    }
  } else if (oper == '5') {
    m_iter = (p1 != 0) ? p2 : m_iter;
  } else if (oper == '6') {
    m_iter = (p1 == 0) ? p2 : m_iter;
  } else if (oper == '7') {
    m_program[out] = (p1 < p2) ? 1 : 0;
  } else if (oper == '8') {
    m_program[out] = (p1 == p2) ? 1 : 0;
  }
}

/* Operator () runs the program until termination and returns what is at
program adress 0 */
tuple<bool, int64_t> Intcode::operator()(int64_t input) {
  return (*this)(std::vector<int64_t>{input});
}

tuple<bool, int64_t> Intcode::operator()(std::vector<int64_t> input) {
  // Append new inputs
  for (auto i : input) {
    m_input.push_back(i);
  }

  // Copy original program incase it needs to be reset
  auto x = m_program;

  while (m_iter < m_program.size()) {
    if (m_program[m_iter] == 99) {
      m_iter = 0, m_base = 0;
      m_program = m_reset ? x : m_program;
      return std::make_tuple(true, m_output);
    }

    // oper, p1, p2, out
    tuple<char, int64_t, int64_t, int64_t> op =
        parse_opcode(std::to_string(m_program[m_iter]));
    char oper = std::get<0>(op);
    int64_t p1 = std::get<1>(op), p2 = std::get<2>(op), out = std::get<3>(op);

    // Temporary output which is nonsense
    if (oper == '3' && m_idle) {
      return std::make_tuple(false, -WINT_MAX);
    }

    operate(oper, p1, p2, out);

    if (!m_reset && oper == '4') {
      return std::make_tuple(false, m_output);
    }
  }
  return std::make_tuple(false, m_output);
}